scipy/stats/_odds_ratio.py

```diff
@@ -0,0 +1,479 @@
+
+from dataclasses import dataclass
+
+import numpy as np
+
+from scipy.special import ndtr, ndtri
+from scipy.optimize import brentq
+from .distributions import nchypergeom_fisher, hypergeom
+from .stats import fisher_exact
+
+
+def _sample_odds_ratio(table):
+    """
+    Given a table [[a, b], [c, d]], compute a*d/(b*c).
+
+    Return nan if the numerator and denominator are 0.
+    Return inf if just the denominator is 0.
+    """
+    # table must be a 2x2 numpy array.
+    if table[1, 0] > 0 and table[0, 1] > 0:
+        oddsratio = table[0, 0] * table[1, 1] / (table[1, 0] * table[0, 1])
+    elif table[0, 0] == 0 or table[1, 1] == 0:
+        oddsratio = np.nan
+    else:
+        oddsratio = np.inf
+    return oddsratio
+
+
+def _solve(func):
+    """
+    Solve func(nc) = 0.  func must be an increasing function.
+    """
+    # We could just as well call the variable `x` instead of `nc`, but we
+    # always call this function with functions for which nc (the noncentrality
+    # parameter) is the variable for which we are solving.
+    nc = 1.0
+    value = func(nc)
+    if value == 0:
+        return nc
+
+    # Multiplicative factor by which to increase or decrease nc when
+    # searching for a bracketing interval.
+    factor = 25.0
+    # Find a bracketing interval.
+    if value > 0:
+        nc /= factor
+        while func(nc) > 0:
+            nc /= factor
+        lo = nc
+        hi = factor*nc
+    else:
+        nc *= factor
+        while func(nc) < 0:
+            nc *= factor
+        lo = nc/factor
+        hi = nc
+
+    # lo and hi bracket the solution for nc.
+    nc = brentq(func, lo, hi, xtol=1e-13)
```

Comment:
Could you comment on why `xtol` is overridden?

Reply:
Initially I suspect I used that value because I've gotten used to other solvers in `scipy.minimize` having fairly large relative tolerances, and I forgot that `brentq` defaults to 2e-12.

In a recent commit, I added some additional tests for the case `kind='sample'`.  Now if I use the default `xtol` in `brentq`, four of those tests fail, and they pass with `xtol=1e-13`.

Reply:
Good enough for me.